### PR TITLE
many: allow updates of kernel and gadget on classic with modes

### PR DESCRIPTION
**Pull request by alfonsosanchezbeato**

### CODE REVIEW

File: boot/boot.go [hunk 1]
```diff
@@ -141,15 +141,35 @@ func Kernel(s snap.PlaceInfo, t snap.Type, dev snap.Device) BootKernel {
 	return trivial{}
 }
 
-func applicable(s snap.PlaceInfo, t snap.Type, dev snap.Device) bool {
-	if dev.Classic() {
+// SnapTypeAffectsBootForDev finds out if a snap type participates in the
+// boot for a given device
+func SnapTypeAffectsBootForDev(t snap.Type, dev snap.Device) bool {
```
Comment (pedronis) [inline boot/boot.go:146]: ForDev is strange, we have other things taking Device here. SnapTypeParticipatesInBoot seems a name more consistent with other things here.

File: overlord/snapstate/handlers.go [hunk 2]
```diff
@@ -1926,21 +1928,39 @@ func setMigrationFlagsinState(snapst *SnapState, snapsup *SnapSetup) {
 	}
 }
 
+// MaybeReboot will schedule a reboot if allowed by the type of system.
+func MaybeReboot(t *state.Task) error {
```
Comment (pedronis) [inline overlord/snapstate/handlers.go:1932]: RestartSystem from here should not be used anymore, so it should be probably be unexported and its logic merged into a MaybeRestartSystem (instead of MaybeReboot) which should live in snapstate.go as MaybeRestartSystem then?  and the core logic (dealing with boot-id etc) of it should move into overlord/restart.MaybeRestartSystem.

File: overlord/snapstate/snapstate.go [hunk 3]
```diff
@@ -710,6 +724,26 @@ var generateSnapdWrappers = backend.GenerateSnapdWrappers
...
+				return err
+			}
+			if currentBootID == changeBootId {
+				task.Logf("Waiting for manual restart...")
+				return &state.Hold{Reason: "waiting for user to reboot"}
+			}
+			logger.Debugf("restart already happened for change %s",
+				task.Change().ID())
+		}
+	}
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go:746]: this should become also some form of overlord/restart helper

File: overlord/state/taskrunner.go [hunk 4]
```diff
@@ -46,6 +46,17 @@ func (r *Retry) Error() string {
 	return "task should be retried"
 }
 
+// Hold is returned from a handler to signal that the task cannot
+// proceed at the moment and that some manual action from the user
+// is usually required at this point.
```
Comment (pedronis) [inline overlord/state/taskrunner.go:51]: this is probably ok but we need to think a bit terminology or comments because so far HoldStatus was meant to thing that could not happen because errors. See the doc comment for it.

File: overlord/stateengine.go [hunk 5]
```diff
@@ -117,6 +122,35 @@ func (se *StateEngine) StartUp() error {
...
+		}
+
+		currentBootID, err := osutil.BootID()
+		if err != nil {
+			return err
+		}
+		if currentBootID != chBootId {
+			logger.Debugf("restart already happened, moving forward change %s", ch.ID())
+			for _, t := range ch.Tasks() {
+				if t.Status() != state.HoldStatus {
```
Comment (pedronis) [inline overlord/stateengine.go:143]: this is too imprecise, there might be other tasks that are in HoldStatus but for other reasons, it seems we need to mark also the task a bit further

File: overlord/stateengine.go [hunk 6]
```diff
@@ -117,6 +122,35 @@ func (se *StateEngine) StartUp() error {
 	return nil
 }
 
+func (se *StateEngine) maybeRestartHeldTasks() error {
```
Comment (pedronis) [inline overlord/stateengine.go:125]: this logic again belongs to the overlord/restart package, maybe it should be a real Manager and then this belongs to its StartUp code

**Review by pedronis — COMMENTED** [review]
thank you for this, I did a first pass, main comments:
* we need to be a bit careful with naming/comments around state.Hold and state.HoldStatus because we are extending a bit the usage patterns of the latter
* we need to mark the relevant tasks with more than just HoldStatus or check for errors or something, as it is we might restart tasks that we shouldn't restart
* all the logic dealing with boot-ids and attaching boot-ids to changes should live in overlord/restart
* we also need to reorg snapstate.RestartSystem that shouldn't be used anymore vs snapstate.MaybeReboot, see comment(s) about that

we also still need to work out what to do about task/change pruning

we also need to work out a splitting strategy

File: boot/boot.go (hunk 1 above)
Comment (alfonsosanchezbeato) [reply boot/boot.go]: Changed

File: overlord/snapstate/handlers.go (hunk 2 above)
Comment (alfonsosanchezbeato) [reply overlord/snapstate/handlers.go]: Done now

File: overlord/snapstate/snapstate.go (hunk 3 above)
Comment (alfonsosanchezbeato) [reply overlord/snapstate/snapstate.go]: Done now

File: overlord/stateengine.go (hunk 5 above)
Comment (alfonsosanchezbeato) [reply overlord/stateengine.go]: Done now

File: overlord/stateengine.go (hunk 6 above)
Comment (alfonsosanchezbeato) [reply overlord/stateengine.go]: Done now

**Comment by alfonsosanchezbeato** [comment]
@pedronis thanks for the review, I've reorganized things a bit. Let me know if you think I should split this, I know this is a bit on the big side but on the other hand is all sort of tightly coupled (also most of the changes is for the spread test).

File: gadget/gadget.go [hunk 7]
```diff
@@ -533,7 +533,7 @@ func classicOrUndetermined(m Model) bool {
 }
 
 func wantsSystemSeed(m Model) bool {
-	return m != nil && m.Grade() != asserts.ModelGradeUnset
+	return m != nil && m.Grade() != asserts.ModelGradeUnset && !m.Classic()
```
Comment (pedronis) [inline gadget/gadget.go:536]: the changes to the gadget package will need their own tests and feel orthogonal to rest, they should be split out

**Review by pedronis — COMMENTED** [review]
did another quick pass, before we start splitting things up. I think the new code in restart is going in the right direction but needs some tweaks and cleaning up because *state packages are either a full manager or have an Init, also it probably doesn't make sense to have both the manager and a separate restartState struct now. In terms of splitting:

* the gadget changes should go in their own PR as we need to think carefully about them and need unit tests
* changes to overlord.go overlord/state and overlord/restart should also be their own PR
* then we should see how big is the rest

**Comment by pedronis** [comment]
closing this for now, some of the code in restart and state will be reused but with other changes in a different PR
